nanoFramework.Device.Bluetooth/BluetoothLEDevice.cs

```diff
@@ -315,36 +324,67 @@ internal GattReadResult ReadAttributeValue(ushort AttributeHandle)
             GattCommunicationStatus status = GattCommunicationStatus.Unreachable;
             Buffer buffer = null;
             ushort result = 0;
+            bool securityStarted = false;
+            bool retryRead;
 
             lock (lockObj)
             {
-                _completedEvent.Reset();
-
-                //Debug.WriteLine($"# ReadValue start con:{ConnectionHandle} attrHandle:{AttributeHandle}  ");
-                result = NativeStartReadValue(ConnectionHandle, AttributeHandle);
-                if (result == 0)
+                do
                 {
-                    // Read started, wait for completed event
-                    if (_completedEvent.WaitOne(operationTimeout, false))
+                    retryRead = false;
+
+                    //Debug.WriteLine($"# ReadAttributeValue");
+
+                    _completedEvent.Reset();
+
+                    //Debug.WriteLine($"# ReadValue start con:{ConnectionHandle} attrHandle:{AttributeHandle}  ");
+                    result = NativeStartReadValue(ConnectionHandle, AttributeHandle);
+                    if (result == 0)
                     {
-                        //Debug.WriteLine($"# ReadValue complete {_eventStatus} ");
-                        // Read completed, check status and get value
-                        if (_eventStatus == 0)
-                        {
-                            buffer = new Buffer(_eventValue);
-                            status = GattCommunicationStatus.Success;
-                        }
-                        else
+                        // Read started, wait for completed event
+                        if (_completedEvent.WaitOne(operationTimeout, false))
                         {
-                            result = _eventStatus;
+                            //Debug.WriteLine($"# ReadValue complete {_eventStatus} ");
+                            // Read completed, check status and get value
+                            if (_eventStatus == 0)
+                            {
+                                buffer = new Buffer(_eventValue);
+                                status = GattCommunicationStatus.Success;
+                            }
+                            else
+                            {
+                                //Debug.WriteLine($"# ReadAttributeValue error {_eventStatus}");
+                                result = _eventStatus;
+                                switch (result)
+                                {
+                                    case 5:
```

Comment:
if result would have been an enum, it would be easier to understand what 5 is

Reply:
add enum